sdk/testing/stepwise/stepwise.go

```diff
@@ -0,0 +1,332 @@
+// a user by creating real resources.
+//
+// Tests will fail unless the verbose flag (`go test -v`, or explicitly
+// the "-test.v" flag) is set. Because some acceptance tests take quite
+// long, we require the verbose flag so users are able to see progress
+// output.
+func Run(tt TestT, c Case) {
+	tt.Helper()
+	// We only run acceptance tests if an env var is set because they're
+	// slow and generally require some outside configuration.
+	checkShouldRun(tt)
+
+	if c.Environment == nil {
+		tt.Fatal("nil driver in acceptance test")
+		// return here only used during testing when using mockT type, otherwise
+		// Fatal will exit
+		return
+	}
+
+	logger := logging.NewVaultLogger(log.Trace)
+
+	if err := c.Environment.Setup(); err != nil {
+		driverErr := fmt.Errorf("error setting up driver: %w", err)
+		if !c.SkipTeardown {
+			if err := c.Environment.Teardown(); err != nil {
+				driverErr = fmt.Errorf("error during driver teardown: %w", driverErr)
+			}
+		}
+		tt.Fatal(driverErr)
+	}
+
+	defer func() {
+		if c.SkipTeardown {
+			logger.Info("driver Teardown skipped")
+			return
+		}
+		if err := c.Environment.Teardown(); err != nil {
+			logger.Error("error in driver teardown:", "error", err)
+		}
+	}()
+
+	// retrieve the root client from the Environment. If this returns an error, fail
+	// immediately
+	rootClient, err := c.Environment.Client()
+	if err != nil {
+		tt.Fatal(err)
+	}
+
+	// Trap the rootToken so that we can preform revocation or other tasks in the
+	// event any steps remove the token during testing.
+	rootToken := c.Environment.RootToken()
+
+	// Defer revocation of any secrets created. We intentionally enclose the
+	// responses slice so in the event of a fatal error during test evaluation, we
+	// are still able to revoke any leases/secrets created
+	var responses []*api.Secret
+	defer func() {
+		// restore root token for admin tasks
+		rootClient.SetToken(rootToken)
+		// failedRevokes tracks any errors we get when attempting to revoke a lease
+		// to log to users at the end of the test.
+		var failedRevokes []*api.Secret
+		for _, secret := range responses {
+			if secret.LeaseID == "" {
+				continue
+			}
+
+			if err := rootClient.Sys().Revoke(secret.LeaseID); err != nil {
+				tt.Error(fmt.Errorf("error revoking lease: %w", err))
+				failedRevokes = append(failedRevokes, secret)
+				continue
+			}
+		}
+
+		// If we have any failed revokes, log it.
+		if len(failedRevokes) > 0 {
+			for _, s := range failedRevokes {
+				tt.Error(fmt.Sprintf(
+					"WARNING: Revoking the following secret failed. It may\n"+
+						"still exist. Please verify:\n\n%#v",
+					s))
+			}
+		}
+	}()
+
+	stepCount := len(c.Steps)
+	for i, step := range c.Steps {
+		// range is zero based, so add 1 for a human friendly output of steps.
+		// "index" here is only used for logging / output, and not to reference the
+		// step in the slice of steps.
+		index := i + 1
```

Comment:
Nit: you could you `i + 1` directly in here and below.

Reply:
Using a variable like `index` gave me a better opportunity to document in code why the `i + 1` what happening at all, so I'd like to keep this unless you feel strongly that it should change

Reply:
Yep, I was suggesting keeping the `range is zero based, so add 1 for a human friendly output of steps` comment and removing the need for the index since that should be clear enough, but not a big deal if it stays.

Reply:
I changed to `i + 1` in 073737336